common/workunit/workunit.cpp

```diff
@@ -2487,8 +2487,74 @@ class CompoundAggregator : implements StatisticAggregator
     IArrayOf<StatisticAggregator> aggregators;
 };
 
+// Conditionally aggregate costs by
+// * excluding hThor costs when excludehThor=true
+// * or hThor costs that are children of workflow that already has a cost
+class FilteredCostAggregator : public CInterfaceOf<IWuScopeVisitor>
+{
+public:
+    FilteredCostAggregator(bool _excludeHThor) : excludeHThor(_excludeHThor) {}
 
-//To calculate aggregates, create a scope iterator, and an instance of a StatisticAggregator, and play the attributes through the interface
+    virtual void noteStatistic(StatisticKind kind, unsigned __int64 value, IConstWUStatistic & extra) override
+    {
+        if (extra.getCreatorType() == SCThthor)
+        {
+            if (excludeHThor)
+                return;
+            if (haveWorkFlowCost)
+            {
+                // Exclude hThor costs that are children of workflow that already has a cost
+                if (isParentScope(workflowScope, extra.queryScope()))
+                    return;
+                else
+                    haveWorkFlowCost = false;
+            }
+        }
+        if (value && extra.getScopeType()==SSTworkflow)
+        {
+            haveWorkFlowCost = true;
+            workflowScope.set(extra.queryScope());
+        }
+        totalCost += value;
+    }
+    virtual void noteAttribute(WuAttr attr, const char * value) override { throwUnexpected(); }
+    virtual void noteHint(const char * kind, const char * value) override { throwUnexpected(); }
+    virtual void noteException(IConstWUException & exception) { throwUnexpected(); }
+    virtual cost_type getTotalCost() const { return totalCost; }
+protected:
+    const bool excludeHThor;
+    bool haveWorkFlowCost = false;
+    StringBuffer workflowScope;
+    cost_type totalCost = 0;
+};
+
+// Aggregrate Thor or hThor costs
+cost_type aggregateCost(const IConstWorkUnit * wu, const char *scope, bool excludeHThor)
+{
+    // depth 1: workflow, depth 2: graph, depth 3: subgraph
+    WuScopeFilter filter;
+    if (scope && *scope) // All costs under specified scope (used to calculate all costs for a workflow)
+    {
+        filter.addScope(scope);
+        filter.setIncludeNesting(3);
+        filter.addOutputStatistic(StCostExecute);
+        filter.addRequiredStat(StCostExecute);
+    }
+    else
+        filter.addFilter("stat[CostExecute],depth[1..3],nested[0],where[CostExecute]");
+    filter.finishedFilter();
+    FilteredCostAggregator aggregator(excludeHThor);
+    Owned<IConstWUScopeIterator> it = &wu->getScopeIterator(filter);
+    for (it->first(); it->isValid(); )
+    {
+        it->playProperties(aggregator);
+        if ((it->getScopeType() == SSTgraph) && (it->queryStat(StCostExecute, MaxStatisticValue) != MaxStatisticValue))
```

Comment:
I'm not sure this should be special cased for a graph.  You can also use getStat to see if the stat exists.

Reply:
If this is not special-cased, then the workflow doesn't include the thor graph costs.